# GG E2 Project

* ist1113837 Guilherme Viana (33%)
  
* ist1114723 André Neves (33%)
  
* ist1113663 Afonso Vinagre (33%)

Prof. Daniel Faria

Lab Shift number: BD25610L-20 and BD25610L-13

## Introdução

Após o diagrama Entidade-Associação para a base de dados “Zoo” ter sido apresentado ao cliente numa primeira entrega, seguiram-se várias iterações para alinhar o desenho da base de dados de acordo com os requisitos do cliente, tendo-se finalmente convergido no esquema SQL apresentado abaixo. 

Pretende-se agora implementar esta base de dados como parte de um sistema de informação que permita a gestão e análise da venda de bilhetes e da popularidade dos vários recintos e animais. A popularidade é um requisito novo, que advém de uma campanha do zoo: cada portador de *bilhete* tem direito a votar no seu *recinto* favorito, o que, na base de dados, incrementa *votos* do *recinto* em 1 e muda *votou* para TRUE no *bilhete*.

`Nota` Optou-se por um sistema de informação separado para a gestão dos funcionários do zoo, que não será abordado no presente trabalho.

## PART I – Base de Dados

#### 0. Criação da Base de Dados

Crie a base de dados `Zoo` no PostgreSQL e execute (nesta base de dados) os comandos para a implementação do respectivo esquema.

In [1]:
%load_ext sql
%config SqlMagic.displaycon = 0
%config SqlMagic.displaylimit = 100
%config SqlMagic.feedback = 0
%sql postgresql+psycopg://app:app@postgres/app --alias zoo

In [10]:
%%sql zoo
DROP TABLE IF EXISTS acesso;

DROP TABLE IF EXISTS bilhete;

DROP TABLE IF EXISTS venda;

DROP TABLE IF EXISTS animal;

DROP TABLE IF EXISTS especie;

DROP TABLE IF EXISTS recinto;

DROP TABLE IF EXISTS zona;

DROP TYPE IF EXISTS cat;

DROP TYPE IF EXISTS cnt;

CREATE TYPE cat AS ENUM(
  'Aves',
  'Carnívoros',
  'Herbívoros',
  'Mamíferos Marinhos',
  'Primatas',
  'Repteis'
);

CREATE TYPE cnt AS ENUM(
  'África',
  'América',
  'Asia',
  'Austrália',
  'Europa'
);

CREATE TABLE zona (
  id_zona SERIAL PRIMARY KEY,
  categoria cat,
  continente cnt,
  preco NUMERIC(4, 2) NOT NULL,
  CONSTRAINT chave_zona UNIQUE (categoria, continente)
);

CREATE TABLE recinto (
  id_recinto SERIAL PRIMARY KEY,
  id_zona INTEGER NOT NULL REFERENCES zona,
  votos INTEGER
);

CREATE TABLE especie (
  nome_cientifico VARCHAR(100) PRIMARY KEY CHECK (nome_cientifico ~ '^[A-Z][a-z]+ [a-z]+$'),
  nome_comum VARCHAR(100) NOT NULL UNIQUE,
  categoria cat NOT NULL,
  continente cnt NOT NULL
);

CREATE TABLE animal (
  id_animal SERIAL PRIMARY KEY,
  nome VARCHAR(80) NOT NULL,
  nome_cientifico VARCHAR(100) NOT NULL REFERENCES especie,
  id_recinto INTEGER NOT NULL REFERENCES recinto,
  data_nasc DATE,
  CONSTRAINT chave_animal UNIQUE (nome, nome_cientifico)
);

CREATE TABLE venda (
  no_venda SERIAL PRIMARY KEY,
  data_hora TIMESTAMP NOT NULL,
  nif_cliente CHAR(9) CHECK (nif_cliente ~ '^[0-9]{9}$')
);

CREATE TABLE bilhete (
  bid SERIAL PRIMARY KEY,
  desconto NUMERIC(4, 2) NOT NULL DEFAULT 0.00,
  votou BOOLEAN NOT NULL DEFAULT FALSE,
  no_venda INTEGER NOT NULL REFERENCES venda
);

CREATE TABLE acesso (
  bid INTEGER NOT NULL REFERENCES bilhete,
  id_zona INTEGER NOT NULL REFERENCES zona,
  PRIMARY KEY (bid, id_zona)
);

RuntimeError: (psycopg.errors.DependentObjectsStillExist) cannot drop table acesso because other objects depend on it
DETAIL:  materialized view vendas_zoo depends on table acesso
HINT:  Use DROP ... CASCADE to drop the dependent objects too.
[SQL: DROP TABLE IF EXISTS acesso;]
(Background on this error at: https://sqlalche.me/e/20/2j85)


Verifique que todas as tabelas foram criadas na base de dados com o seguinte comando:

In [7]:
%sqlcmd tables

Name
especie
animal
zona
acesso
bilhete
venda
recinto


#### 1. Restrições de Integridade [4 valores]

Recorrendo a `Triggers apenas quando estritamente necessário`, implemente na base de dados “Zoo” as seguintes restrições de integridade:


1. `RI-1` Em zona, a categoria e o continente não podem ser simultaneamente `NULL`
* uma zona é sempre dedicada a uma categoria de animais, a um continente, ou a ambos.

In [8]:
%%sql zoo
ALTER TABLE zona
ADD CONSTRAINT RI1_zona CHECK (
  categoria IS NOT NULL
  OR continente IS NOT NULL
)

RuntimeError: (psycopg.errors.DuplicateObject) constraint "ri1_zona" for relation "zona" already exists
[SQL: ALTER TABLE zona
ADD CONSTRAINT RI1_zona CHECK (
  categoria IS NOT NULL
  OR continente IS NOT NULL
)]
(Background on this error at: https://sqlalche.me/e/20/f405)


2. `RI-2` Um animal tem de estar alojado num recinto que esteja numa zona compatível
* se a zona tiver categoria definida, tem de corresponder à categoria da espécie do animal;
* se tiver continente definido, tem de corresponder ao continente da espécie do animal.

In [5]:
%%sql zoo
CREATE OR REPLACE FUNCTION check_compatibility () RETURNS TRIGGER AS $$
BEGIN
    IF EXISTS(
        SELECT 1 
        FROM recinto r
        JOIN zona z ON r.id_zona = z.id_zona
        JOIN especie e ON e.nome_cientifico= NEW.nome_cientifico
        WHERE r.id_recinto = NEW.id_recinto
        AND (
              (z.categoria IS NOT NULL AND z.categoria != e.categoria)
              OR 
              (z.continente IS NOT NULL AND z.continente != e.continente)
              )
        )THEN 
            RAISE EXCEPTION 'Recinto Incompativel com a zona da especie';
    END IF;
    RETURN NEW;
END;
$$ LANGUAGE plpgsql;

DROP TRIGGER IF EXISTS ri2_animal_trigger ON animal;

CREATE TRIGGER ri2_animal_trigger BEFORE INSERT
OR
UPDATE ON animal FOR EACH ROW
EXECUTE FUNCTION check_compatibility ();

++
||
++
++

`RI-3` Todos os animais de uma espécie têm de estar alojados em recinto(s) de uma mesma zona
* `Nota` No entanto, podem existir várias zonas compatíveis com a espécie.

In [6]:
%%sql zoo
CREATE OR REPLACE FUNCTION check_valid_place () RETURNS TRIGGER AS $$
BEGIN
    IF EXISTS(
        SELECT 1
        FROM animal a
        JOIN recinto r1 ON r1.id_recinto = NEW.id_recinto
        JOIN recinto r2 ON r2.id_recinto = a.id_recinto
        WHERE a.nome_cientifico= NEW.nome_cientifico
        AND r1.id_zona!= r2.id_zona
    )THEN 
        RAISE EXCEPTION 'Zona do Recinto não é compativel';
    END IF;
    RETURN NEW;
END;
$$ LANGUAGE plpgsql;

DROP TRIGGER IF EXISTS ri3_animal_recinto_trigger ON animal;

CREATE TRIGGER ri3_animal_recinto_trigger BEFORE INSERT
OR
UPDATE ON animal FOR EACH ROW
EXECUTE FUNCTION check_valid_place ();

++
||
++
++

`RI-4` Após uma transação de venda
* uma venda tem de incluir pelo menos um bilhete com acesso a pelo menos uma zona do zoo.

In [7]:
%%sql zoo
CREATE OR REPLACE FUNCTION check_valid_sell () RETURNS TRIGGER AS $$
BEGIN
    IF NOT EXISTS(
        SELECT 1
            FROM acesso a
            JOIN bilhete b ON b.bid = a.bid
            WHERE b.no_venda = NEW.no_venda
    )THEN 
        RAISE EXCEPTION 'Venda Inválida';
    END IF;
    RETURN NEW;
END;
$$ LANGUAGE plpgsql;

DROP TRIGGER IF EXISTS ri4_invalid_sell ON venda;

CREATE CONSTRAINT TRIGGER ri4_invalid_sell
AFTER INSERT
OR
UPDATE ON venda DEFERRABLE INITIALLY DEFERRED FOR EACH ROW
EXECUTE FUNCTION check_valid_sell ();

++
||
++
++

#### 2. Preenchimento da Base de Dados [3 valores]

##### Critérios de Avaliação

* Cumprimento dos requisitos de cobertura  
* Resultado não vazio em todas as consultas do ponto 5

`Nota`: Caso não consiga implementar alguma(s) das restrições de integridade do ponto 1, deve ainda assim assegurar-se que o preenchimento da base dados as cumpre. 

Pode utilizar ferramentas de **IA generativo**, scripts ou qualquer outro meio para gerar os comandos INSERT. Preencha todas as tabelas da base de dados de forma consistente, após execução do ponto anterior para garantir que são respeitadas todas as restrições de integridade ali expressas, e com os seguintes **requisitos de cobertura** adicionais:

1. Têm de haver pelo menos 7 ***zonas***, das quais:
* O preço de acesso a cada zona deve estar entre 5€ e 30€.
* pelo menos 3 são dedicadas exclusivamente a uma *categoria* (sem *continente* preenchido);
* pelo menos 2 são dedicadas exclusivamente a um *continente* (sem *categoria* preenchida);
* pelo menos 2 têm *categoria* e *continente* preenchidos.

`Nota` As 2 ou mais zonas que têm *categoria* e *continente* preenchidos têm de partilhar entre todas elas um único valor de um desses dois atributos (sendo naturalmente o outro diferente, devido à restrição de chave candidata da tabela), e o valor desse atributo não pode ocorrer sozinho (i.e., qualquer zona que tenha o valor desse atributo não pode ter o outro atributo a NULL). O valor desse atributo corresponde à “especialidade” do zoo.  
* `Exemplo 1` Um zoo pode ter como especialidade herbívoros, e nesse caso terá de ter pelo menos 2 zonas com categoria “herbívoro” e continentes diferentes, não podendo ter nenhuma zona com categoria “herbívoro” sem continente definido. Terá adicionalmente pelo menos 3 zonas dedicadas a outras categorias que não “herbívoro” e 2 zonas dedicadas a continentes.  
* `Exemplo 2` Outro zoo pode ter como especialidade animais de África, e nesse caso terá de ter pelo menos 2 zonas com continente “África” e categorias diferentes, não podendo ter nenhuma zona com continente “África” sem categoria definida. Terá adicionalmente pelo menos 3 zonas dedicadas a categorias e 2 zonas dedicadas a continentes que não “África”.


In [8]:
%%sql zoo
INSERT INTO
  zona (id_zona, categoria, continente, preco)
VALUES
  (1, 'Carnívoros', 'África', 15.00);

INSERT INTO
  zona (id_zona, categoria, continente, preco)
VALUES
  (2, 'Herbívoros', 'África', 12.00);

INSERT INTO
  zona (id_zona, categoria, continente, preco)
VALUES
  (3, 'Aves', NULL, 8.00);

INSERT INTO
  zona (id_zona, categoria, continente, preco)
VALUES
  (4, 'Primatas', NULL, 10.00);

INSERT INTO
  zona (id_zona, categoria, continente, preco)
VALUES
  (5, 'Repteis', NULL, 6.00);

INSERT INTO
  zona (id_zona, categoria, continente, preco)
VALUES
  (6, NULL, 'Europa', 20.00);

INSERT INTO
  zona (id_zona, categoria, continente, preco)
VALUES
  (7, NULL, 'Asia', 25.00);

++
||
++
++

2. Cada *zona* tem de ter entre 10 e 30 ***recintos***, e cada recinto deve ter no mínimo 0.1% dos votos totais (ver 5. abaixo). Os votos devem refletir os *acessos* dos *bilhetes*:
* O somatório global dos *votos* de todos os ***recintos*** tem de ser igual ao número de ***bilhetes*** vendidos com *votou* TRUE; 
* O somatório dos *votos* de todos os ***recintos*** de uma ***zona*** tem de ser menor ou igual ao número de ***bilhetes*** vendidos com acesso a essa ***zona*** e *votou* TRUE.

`Nota`: Pode preencher os votos dos ***recintos*** já no INSERT destes, ou alternativamente após o preenchimento dos *bilhetes*, por meio de comandos de UPDATE.

In [9]:
%%sql zoo
INSERT INTO
  recinto (id_zona, votos)
SELECT
  id_zona,
  0
FROM
  generate_series(1, 7) AS id_zona,
  generate_series(1, 15) AS n_recinto;

++
||
++
++

3. Têm de haver pelo menos 200 ***espécies*** reais cobrindo todas as *categorias* e todos os *continentes*.  

In [31]:
%%sql zoo
INSERT INTO especie (nome_cientifico, nome_comum, categoria, continente) VALUES 
-- CARNÍVOROS - ÁFRICA (Zona 1)
('Panthera leo', 'Leão', 'Carnívoros', 'África'),
('Panthera pardus', 'Leopardo', 'Carnívoros', 'África'),
('Acinonyx jubatus', 'Chita', 'Carnívoros', 'África'),
('Crocuta crocuta', 'Hiena malhada', 'Carnívoros', 'África'),
('Lycaon pictus', 'Cão selvagem', 'Carnívoros', 'África'),
('Caracal caracal', 'Caracal', 'Carnívoros', 'África'),
('Leptailurus serval', 'Serval', 'Carnívoros', 'África'),
('Suricata suricatta', 'Suricata', 'Carnívoros', 'África'),
('Canis mesomelas', 'Chacal de dorso negro', 'Carnívoros', 'África'),
('Mellivora capensis', 'Texugo do mel', 'Carnívoros', 'África'),
('Hyaena hyaena', 'Hiena riscada', 'Carnívoros', 'África'),
('Parahyaena brunnea', 'Hiena castanha', 'Carnívoros', 'África'),
('Proteles cristata', 'Lobo da terra', 'Carnívoros', 'África'),
('Felis lybica', 'Gato selvagem africano', 'Carnívoros', 'África'),
('Felis nigripes', 'Gato de patas negras', 'Carnívoros', 'África'),
('Caracal aurata', 'Gato dourado africano', 'Carnívoros', 'África'),
('Otocyon megalotis', 'Raposa orelhuda', 'Carnívoros', 'África'),
('Vulpes zerda', 'Feneco', 'Carnívoros', 'África'),
('Vulpes chama', 'Raposa do cabo', 'Carnívoros', 'África'),
('Canis lupaster', 'Lobo dourado africano', 'Carnívoros', 'África'),
('Canis adustus', 'Chacal listrado', 'Carnívoros', 'África'),
('Civettictis civetta', 'Civeta africana', 'Carnívoros', 'África'),
('Genetta genetta', 'Gineta comum', 'Carnívoros', 'África'),
('Herpestes ichneumon', 'Mangusto egípcio', 'Carnívoros', 'África'),
('Mungos mungo', 'Mangusto listrado', 'Carnívoros', 'África'),
('Cynictis penicillata', 'Mangusto amarelo', 'Carnívoros', 'África'),
('Galerella sanguinea', 'Mangusto esguio', 'Carnívoros', 'África'),
('Ichneumia albicauda', 'Mangusto de cauda branca', 'Carnívoros', 'África'),
('Atilax paludinosus', 'Mangusto do pântano', 'Carnívoros', 'África'),
('Nandinia binotata', 'Civeta de palmeira', 'Carnívoros', 'África'),
-- HERBÍVOROS - ÁFRICA (Zona 2)
('Giraffa camelopardalis', 'Girafa', 'Herbívoros', 'África'),
('Loxodonta africana', 'Elefante africano', 'Herbívoros', 'África'),
('Loxodonta cyclotis', 'Elefante da floresta', 'Herbívoros', 'África'),
('Hippopotamus amphibius', 'Hipopótamo', 'Herbívoros', 'África'),
('Choeropsis liberiensis', 'Hipopótamo pigmeu', 'Herbívoros', 'África'),
('Equus quagga', 'Zebra da planície', 'Herbívoros', 'África'),
('Equus grevyi', 'Zebra de grevy', 'Herbívoros', 'África'),
('Syncerus caffer', 'Búfalo africano', 'Herbívoros', 'África'),
('Diceros bicornis', 'Rinoceronte negro', 'Herbívoros', 'África'),
('Ceratotherium simum', 'Rinoceronte branco', 'Herbívoros', 'África'),
('Tragelaphus strepsiceros', 'Cudo', 'Herbívoros', 'África'),
('Connochaetes taurinus', 'Gnu', 'Herbívoros', 'África'),
('Oryx gazella', 'Orix', 'Herbívoros', 'África'),
('Phacochoerus africanus', 'Facocero', 'Herbívoros', 'África'),
('Potamochoerus larvatus', 'Porco do mato', 'Herbívoros', 'África'),
('Aepyceros melampus', 'Impala', 'Herbívoros', 'África'),
('Eudorcas thomsonii', 'Gazela de thomson', 'Herbívoros', 'África'),
('Tragelaphus eurycerus', 'Bongo', 'Herbívoros', 'África'),
('Taurotragus oryx', 'Elande', 'Herbívoros', 'África'),
('Alcelaphus buselaphus', 'Búbalo', 'Herbívoros', 'África'),
('Hippotragus niger', 'Palanca negra', 'Herbívoros', 'África'),
('Kobus ellipsiprymnus', 'Inhacoso', 'Herbívoros', 'África'),
('Damaliscus lunatus', 'Topi', 'Herbívoros', 'África'),
('Okapia johnstoni', 'Ocapí', 'Herbívoros', 'África'),
('Redunca redunca', 'Redunca', 'Herbívoros', 'África'),
('Sylvicapra grimmia', 'Cabrito do mato', 'Herbívoros', 'África'),
('Tragelaphus angasii', 'Niala', 'Herbívoros', 'África'),
('Litocranius walleri', 'Gerenuque', 'Herbívoros', 'África'),
('Antidorcas marsupialis', 'Cabra de leque', 'Herbívoros', 'África'),
('Oreotragus oreotragus', 'Cabrito das pedras', 'Herbívoros', 'África'),
-- AVES - MUNDO (Zona 3)
('Haliaeetus leucocephalus', 'Águia careca', 'Aves', 'América'),
('Spheniscus magellanicus', 'Pinguim de magalhães', 'Aves', 'América'),
('Cyanocitta cristata', 'Gaio azul', 'Aves', 'América'),
('Cardinalis cardinalis', 'Cardeal', 'Aves', 'América'),
('Eudocimus ruber', 'Íbis escarlate', 'Aves', 'América'),
('Pandion haliaetus', 'Águia pesqueira', 'Aves', 'Europa'),
('Alcedo atthis', 'Guarda rios', 'Aves', 'Europa'),
('Tyto alba', 'Coruja das torres', 'Aves', 'Europa'),
('Falco peregrinus', 'Falcão peregrino', 'Aves', 'Europa'),
('Gyps fulvus', 'Grifo', 'Aves', 'Europa'),
('Phalacrocorax carbo', 'Corvo marinho', 'Aves', 'Europa'),
('Pelecanus onocrotalus', 'Pelicano branco', 'Aves', 'África'),
('Bucorvus leadbeateri', 'Calau terrestre', 'Aves', 'África'),
('Torgos tracheliotos', 'Abutre real', 'Aves', 'África'),
('Spheniscus demersus', 'Pinguim africano', 'Aves', 'África'),
('Strigops habroptila', 'Kakapo', 'Aves', 'Austrália'),
('Apteryx australis', 'Kiwi', 'Aves', 'Austrália'),
('Goura victoria', 'Pombo coroado', 'Aves', 'Austrália'),
('Cacatua galerita', 'Cacatua de crista amarela', 'Aves', 'Austrália'),
('Paradisaea apoda', 'Ave do paraíso', 'Aves', 'Austrália'),
('Fregata magnificens', 'Fragata', 'Aves', 'América'),
('Ramphastos sulfuratus', 'Tucano de bico verde', 'Aves', 'América'),
('Buteo jamaicensis', 'Águia de cauda vermelha', 'Aves', 'América'),
('Sula nebouxii', 'Patola de pés azuis', 'Aves', 'América'),
('Aquila chrysaetos', 'Águia real', 'Aves', 'Europa'),
('Ciconia nigra', 'Cegonha preta', 'Aves', 'Europa'),
('Garrulus glandarius', 'Gaio comum', 'Aves', 'Europa'),
('Dacelo leachii', 'Cucaburra de asas azuis', 'Aves', 'Austrália'),
('Pavo muticus', 'Pavão verde', 'Aves', 'Asia'),
('Aethopyga siparaja', 'Beija flor asiático', 'Aves', 'Asia'),
-- PRIMATAS - MUNDO (Zona 4)
('Lemur macaco', 'Lémure negro', 'Primatas', 'África'),
('Microcebus murinus', 'Lémure rato', 'Primatas', 'África'),
('Daubentonia madagascariensis', 'Aie aie', 'Primatas', 'África'),
('Propithecus verreauxi', 'Sifaca', 'Primatas', 'África'),
('Indri indri', 'Indri', 'Primatas', 'África'),
('Varecia variegata', 'Lémure de colar', 'Primatas', 'África'),
('Erythrocebus patas', 'Macaco patas', 'Primatas', 'África'),
('Theropithecus gelada', 'Gelada', 'Primatas', 'África'),
('Pan paniscus', 'Bonobo', 'Primatas', 'África'),
('Gorilla beringei', 'Gorila da montanha', 'Primatas', 'África'),
('Cercocebus atys', 'Mangabey fuliginoso', 'Primatas', 'África'),
('Lophocebus albigena', 'Mangabey de bochechas cinzentas', 'Primatas', 'África'),
('Miopithecus talapoin', 'Talapoin', 'Primatas', 'África'),
('Galago senegalensis', 'Gálago do senegal', 'Primatas', 'África'),
('Perodicticus potto', 'Potto', 'Primatas', 'África'),
('Alouatta palliata', 'Bugio de manto', 'Primatas', 'América'),
('Ateles geoffroyi', 'Macaco aranha de geoffroy', 'Primatas', 'América'),
('Cebus apella', 'Macaco prego de crista', 'Primatas', 'América'),
('Saimiri oerstedii', 'Macaco esquilo de dorso vermelho', 'Primatas', 'América'),
('Aotus trivirgatus', 'Macaco da noite', 'Primatas', 'América'),
('Callithrix jacchus', 'Sagui de tufo branco', 'Primatas', 'América'),
('Cebuella pygmaea', 'Sagui pigmeu', 'Primatas', 'América'),
('Pithecia pithecia', 'Parauacu', 'Primatas', 'América'),
('Callicebus moloch', 'Zogue zogue', 'Primatas', 'América'),
('Brachyteles arachnoides', 'Muriqui', 'Primatas', 'América'),
('Pongo abelii', 'Orangotango de sumatra', 'Primatas', 'Asia'),
('Symphalangus syndactylus', 'Siamang', 'Primatas', 'Asia'),
('Nomascus leucogenys', 'Gibão de bochechas brancas', 'Primatas', 'Asia'),
('Macaca fascicularis', 'Macaco caranguejeiro', 'Primatas', 'Asia'),
('Tarsius syrichta', 'Társio das filipinas', 'Primatas', 'Asia'),
-- RÉPTEIS - MUNDO (Zona 5)
('Crocodylus porosus', 'Crocodilo de água salgada', 'Repteis', 'Asia'),
('Alligator sinensis', 'Jacaré da china', 'Repteis', 'Asia'),
('Caiman crocodilus', 'Caimão de lunetas', 'Repteis', 'América'),
('Melanosuchus niger', 'Jacaré açu', 'Repteis', 'América'),
('Paleosuchus palpebrosus', 'Jacaré anão', 'Repteis', 'América'),
('Sphenodon punctatus', 'Tuatara', 'Repteis', 'Austrália'),
('Heloderma suspectum', 'Monstro de gila', 'Repteis', 'América'),
('Heloderma horridum', 'Lagarto de contas', 'Repteis', 'América'),
('Cyclura cornuta', 'Iguana rinoceronte', 'Repteis', 'América'),
('Amblyrhynchus cristatus', 'Iguana marinha', 'Repteis', 'América'),
('Conolophus subcristatus', 'Iguana terrestre de galápagos', 'Repteis', 'América'),
('Basiliscus plumifrons', 'Basilisco verde', 'Repteis', 'América'),
('Pantherophis guttatus', 'Cobra do milho', 'Repteis', 'América'),
('Lampropeltis triangulum', 'Cobra do leite', 'Repteis', 'América'),
('Eunectes notaeus', 'Sucuri amarela', 'Repteis', 'América'),
('Epicrates cenchria', 'Jiboia arco íris', 'Repteis', 'América'),
('Python regius', 'Píton real', 'Repteis', 'África'),
('Malayopython reticulatus', 'Píton reticulada', 'Repteis', 'Asia'),
('Naja haje', 'Naja egípcia', 'Repteis', 'África'),
('Crotalus durissus', 'Cascavel tropical', 'Repteis', 'América'),
('Agkistrodon contortrix', 'Boca de cobre', 'Repteis', 'América'),
('Agkistrodon piscivorus', 'Mocassim de água', 'Repteis', 'América'),
('Lachesis muta', 'Surucucu', 'Repteis', 'América'),
('Bothrops jararaca', 'Jararaca', 'Repteis', 'América'),
('Chelonoidis nigra', 'Tartaruga gigante de galápagos', 'Repteis', 'América'),
('Aldabrachelys gigantea', 'Tartaruga gigante de aldabra', 'Repteis', 'África'),
('Dermochelys coriacea', 'Tartaruga de couro', 'Repteis', 'América'),
('Chelonia mydas', 'Tartaruga verde', 'Repteis', 'América'),
('Eretmochelys imbricata', 'Tartaruga de pente', 'Repteis', 'América'),
('Macrochelys temminckii', 'Tartaruga aligátor', 'Repteis', 'América'),
-- MAMÍFEROS - EUROPA (Zona 6)
('Lynx lynx', 'Lince eurasiático', 'Carnívoros', 'Europa'),
('Canis aureus', 'Chacal dourado', 'Carnívoros', 'Europa'),
('Nyctereutes procyonoides', 'Cão guaxinim', 'Carnívoros', 'Europa'),
('Vulpes lagopus', 'Raposa do ártico', 'Carnívoros', 'Europa'),
('Mustela nivalis', 'Doninha', 'Carnívoros', 'Europa'),
('Mustela lutreola', 'Vison europeu', 'Carnívoros', 'Europa'),
('Vormela peregusna', 'Tourão marmoreado', 'Carnívoros', 'Europa'),
('Phoca vitulina', 'Foca comum', 'Carnívoros', 'Europa'),
('Halichoerus grypus', 'Foca cinzenta', 'Carnívoros', 'Europa'),
('Monachus monachus', 'Foca monge do mediterrâneo', 'Carnívoros', 'Europa'),
('Odobenus rosmarus', 'Morsa', 'Carnívoros', 'Europa'),
('Sorex araneus', 'Musaranho comum', 'Carnívoros', 'Europa'),
('Neomys fodiens', 'Musaranho de água', 'Carnívoros', 'Europa'),
('Alces alces', 'Alce', 'Herbívoros', 'Europa'),
('Rangifer tarandus', 'Rena', 'Herbívoros', 'Europa'),
('Capra ibex', 'Íbex alpino', 'Herbívoros', 'Europa'),
('Capra pyrenaica', 'Íbex ibérico', 'Herbívoros', 'Europa'),
('Rupicapra pyrenaica', 'Camurça dos pirinéus', 'Herbívoros', 'Europa'),
('Saiga tatarica', 'Saiga', 'Herbívoros', 'Europa'),
('Castor fiber', 'Castor europeu', 'Herbívoros', 'Europa'),
('Lepus europaeus', 'Lebre europeia', 'Herbívoros', 'Europa'),
('Lepus timidus', 'Lebre da montanha', 'Herbívoros', 'Europa'),
('Oryctolagus cuniculus', 'Coelho bravo', 'Herbívoros', 'Europa'),
('Cricetus cricetus', 'Hamster europeu', 'Herbívoros', 'Europa'),
('Lemmus lemmus', 'Lêmingue', 'Herbívoros', 'Europa'),
('Glis glis', 'Lírio', 'Herbívoros', 'Europa'),
('Muscardinus avellanarius', 'Arganaz', 'Herbívoros', 'Europa'),
('Eliomys quercinus', 'Leirão', 'Herbívoros', 'Europa'),
('Marmota bobak', 'Marmota bobak', 'Herbívoros', 'Europa'),
('Hystrix cristata', 'Porco espinho com crista', 'Herbívoros', 'Europa'),
-- MAMÍFEROS - ÁSIA (Zona 7)
('Ursus thibetanus', 'Urso negro asiático', 'Carnívoros', 'Asia'),
('Catopuma temminckii', 'Gato bravo dourado', 'Carnívoros', 'Asia'),
('Prionailurus viverrinus', 'Gato pescador', 'Carnívoros', 'Asia'),
('Arctictis binturong', 'Binturong', 'Carnívoros', 'Asia'),
('Vulpes corsac', 'Raposa corsac', 'Carnívoros', 'Asia'),
('Martes flavigula', 'Marta de garganta amarela', 'Carnívoros', 'Asia'),
('Mustela sibirica', 'Doninha siberiana', 'Carnívoros', 'Asia'),
('Paguma larvata', 'Civeta de máscara', 'Carnívoros', 'Asia'),
('Paradoxurus hermaphroditus', 'Civeta de palmeira comum', 'Carnívoros', 'Asia'),
('Ailurus fulgens', 'Panda vermelho', 'Herbívoros', 'Asia'),
('Muntiacus muntjak', 'Muntíaco', 'Herbívoros', 'Asia'),
('Bos mutus', 'Iaque', 'Herbívoros', 'Asia'),
('Budorcas taxicolor', 'Takin', 'Herbívoros', 'Asia'),
('Pseudois nayaur', 'Carneiro azul', 'Herbívoros', 'Asia'),
('Hemitragus jemlahicus', 'Tahr dos himalaias', 'Herbívoros', 'Asia'),
('Equus hemionus', 'Hemíono', 'Herbívoros', 'Asia'),
('Equus kiang', 'Kiang', 'Herbívoros', 'Asia'),
('Procapra gutturosa', 'Gazela da mongólia', 'Herbívoros', 'Asia'),
('Naemorhedus goral', 'Goral', 'Herbívoros', 'Asia'),
('Axis porcinus', 'Cervo porco', 'Herbívoros', 'Asia');

++
||
++
++

4. O número de ***animais*** de cada *espécie* deve ser entre 1 e 12, com média entre 2 e 3 animais
* Um animal só pode estar num *recinto* sozinho se há 3 ou menos ***animais*** dessa *espécie*.
* Têm de haver no zoo alguns *recintos* com:
    * apenas um ***animal***;
    * vários ***animais*** de uma só *espécie*;
    * vários ***animais*** de várias *espécies*.

In [32]:
%%sql zoo
DO $$
DECLARE
    r_esp RECORD;
    v_zona_id INT;
    v_recinto_id INT;
    v_qtd_animais INT;
    v_rand NUMERIC;
BEGIN
    -- 1. Limpar a tabela de animais para não duplicar dados
    DELETE FROM animal;

    -- 2. DISTRIBUIÇÃO DOS ANIMAIS
    FOR r_esp IN SELECT * FROM especie LOOP

        -- Gerar UM único número aleatório para esta espécie
        v_rand := random();

        -- A. Matemática Exata (Média garantida entre 2.0 e 3.0)
        IF v_rand < 0.45 THEN
            v_qtd_animais := 1;
        ELSIF v_rand < 0.75 THEN  
            v_qtd_animais := 2;
        ELSIF v_rand < 0.85 THEN  
            v_qtd_animais := 3;
        ELSE
            v_qtd_animais := floor(random() * 9) + 4; 
        END IF;

        -- B. Encontrar a zona compatível para a espécie
        SELECT z.id_zona INTO v_zona_id
        FROM zona z
        WHERE (z.categoria IS NULL OR z.categoria = r_esp.categoria)
          AND (z.continente IS NULL OR z.continente = r_esp.continente)
        LIMIT 1;

        -- C. Respeitar a infraestrutura: Só insere se encontrar uma zona válida
        IF v_zona_id IS NOT NULL THEN
            
            -- Inteligência de Alocação de Recintos
            IF v_qtd_animais = 1 THEN
                SELECT id_recinto INTO v_recinto_id FROM recinto r
                WHERE id_zona = v_zona_id
                  AND NOT EXISTS (SELECT 1 FROM animal a WHERE a.id_recinto = r.id_recinto)
                LIMIT 1;

                IF v_recinto_id IS NULL THEN
                    SELECT id_recinto INTO v_recinto_id FROM recinto WHERE id_zona = v_zona_id ORDER BY random() LIMIT 1;
                END IF;
            ELSE
                SELECT id_recinto INTO v_recinto_id FROM recinto r
                WHERE id_zona = v_zona_id
                  AND (
                       (SELECT count(*) FROM animal a WHERE a.id_recinto = r.id_recinto) != 1
                       OR NOT EXISTS (SELECT 1 FROM animal a WHERE a.id_recinto = r.id_recinto)
                  )
                ORDER BY random() LIMIT 1;

                IF v_recinto_id IS NULL THEN
                    SELECT id_recinto INTO v_recinto_id FROM recinto WHERE id_zona = v_zona_id ORDER BY random() LIMIT 1;
                END IF;
            END IF;

            -- D. Inserir animais na base de dados
            FOR i IN 1..v_qtd_animais LOOP
                INSERT INTO animal (nome, nome_cientifico, id_recinto, data_nasc)
                VALUES (
                    r_esp.nome_comum || ' ' || i,
                    r_esp.nome_cientifico,
                    v_recinto_id,
                    CAST('2015-01-01' AS DATE) + CAST(floor(random() * 3000) AS INT)
                );
        
            END LOOP;
            
        END IF;

    END LOOP;
END $$;

++
||
++
++

5. Têm de haver ***vendas*** de ***bilhetes*** para todos os dias de 2026 até 11 de Junho (i.e., `[2026-01-01 - 2026-06-11]`), tais que:
* Haja pelo menos 1000 ***bilhetes*** nos dias de semana;
* Haja pelo menos 4000 ***bilhetes*** nos fins de semana;
* Cerca de metade dos ***bilhetes*** por dia sejam para crianças ou séniores, tendo 50% de *desconto*, e os restantes não têm *desconto*;
* Pelo menos 2% dos ***bilhetes*** de cada dia tenham `Acesso Total` (i.e., ***acesso*** a todas as *zonas*);
* Todas as ***zonas*** estejam em pelo menos 10 ***bilhetes*** por dia;
* Cada ***bilhete*** tenha ***acesso*** a 3 ou mais ***zonas***, e cada ***zona*** tem de figurar em pelo menos 25% dos ***bilhetes***;
* Tem de haver ***bilhetes*** para todas as combinações de 3 ou mais ***zonas*** (mas não necessariamente todos os dias);
* Pelo menos 75% dos ***bilhetes*** têm *votou* a TRUE.

`Nota` A implementação da RI-4 no ponto anterior obriga a que a inserção nestas três tabelas seja encapsulada numa transação. 

In [12]:
import psycopg
from datetime import date, timedelta

conn = psycopg.connect("host=postgres dbname=app user=app password=app", autocommit=False)

start = date(2026, 1, 1)
end   = date(2026, 6, 11)
days  = [start + timedelta(days=i) for i in range((end - start).days + 1)]

with conn.cursor() as cur:
    cur.execute("TRUNCATE acesso, bilhete, venda RESTART IDENTITY CASCADE")
    conn.commit()
    
    total_zonas = cur.execute("SELECT count(*) FROM zona").fetchone()[0]
    print(f"✓ Tabelas limpas | {total_zonas} zonas", flush=True)

    for d in days:
        qt = 4000 if d.isoweekday() >= 6 else 1000

        cur.execute("DROP TABLE IF EXISTS tmp_venda")
        cur.execute("""
            CREATE TEMP TABLE tmp_venda AS
            SELECT
                nextval(pg_get_serial_sequence('venda', 'no_venda')) AS no_venda,
                CAST(%s + (random() * '9 hours'::interval) + '9 hours'::interval AS TIMESTAMP) AS data_hora,
                LPAD(CAST(floor(random() * 999999999) + 1 AS VARCHAR), 9, '0') AS nif_cliente
            FROM generate_series(1, %s)
        """, (d, qt))

        cur.execute("INSERT INTO venda (no_venda, data_hora, nif_cliente) SELECT no_venda, data_hora, nif_cliente FROM tmp_venda")

        cur.execute("DROP TABLE IF EXISTS tmp_bilhete")
        cur.execute("""
            CREATE TEMP TABLE tmp_bilhete AS
            SELECT
                nextval(pg_get_serial_sequence('bilhete', 'bid')) AS bid,
                no_venda,
                CASE WHEN random() < 0.5 THEN 50.00 ELSE 0.00 END AS desconto,
                (random() < 0.80) AS votou
            FROM tmp_venda
        """)

        cur.execute("INSERT INTO bilhete (bid, desconto, votou, no_venda) SELECT bid, desconto, votou, no_venda FROM tmp_bilhete")

        # 2% dos bilhetes com acesso total (todas as zonas)
        cur.execute("""
            INSERT INTO acesso (bid, id_zona)
            SELECT b.bid, z.id_zona
            FROM tmp_bilhete b
            CROSS JOIN zona z
            WHERE b.bid % 50 = 0
        """)

        # Restantes com 3 a total_zonas zonas aleatórias
        cur.execute("""
            INSERT INTO acesso (bid, id_zona)
            SELECT DISTINCT b.bid, z.id_zona
            FROM tmp_bilhete b
            CROSS JOIN LATERAL (
                SELECT id_zona FROM zona
                ORDER BY random()
                LIMIT CAST(floor(random() * (%s - 2)) + 3 AS INT)
            ) z
            WHERE b.bid %% 50 != 0
            ON CONFLICT DO NOTHING
        """, (total_zonas,))

        conn.commit()
        print(f"✓ {d} ({qt} bilhetes)", flush=True)

    cur.execute("""
        UPDATE recinto SET votos = floor(
            (SELECT count(*) FROM bilhete WHERE votou)::numeric /
            (SELECT count(*) FROM recinto)
        )
    """)
    cur.execute("""
        UPDATE recinto SET votos = votos + mod(
            (SELECT count(*) FROM bilhete WHERE votou)::int,
            (SELECT count(*) FROM recinto)::int
        )
        WHERE id_recinto = (SELECT min(id_recinto) FROM recinto)
    """)
    conn.commit()
    print("✓ Votos atualizados", flush=True)

conn.close()
print("🎉 Concluído!")

✓ Tabelas limpas | 7 zonas
✓ 2026-01-01 (1000 bilhetes)
✓ 2026-01-02 (1000 bilhetes)
✓ 2026-01-03 (4000 bilhetes)
✓ 2026-01-04 (4000 bilhetes)
✓ 2026-01-05 (1000 bilhetes)
✓ 2026-01-06 (1000 bilhetes)
✓ 2026-01-07 (1000 bilhetes)
✓ 2026-01-08 (1000 bilhetes)
✓ 2026-01-09 (1000 bilhetes)
✓ 2026-01-10 (4000 bilhetes)
✓ 2026-01-11 (4000 bilhetes)
✓ 2026-01-12 (1000 bilhetes)
✓ 2026-01-13 (1000 bilhetes)
✓ 2026-01-14 (1000 bilhetes)
✓ 2026-01-15 (1000 bilhetes)
✓ 2026-01-16 (1000 bilhetes)
✓ 2026-01-17 (4000 bilhetes)
✓ 2026-01-18 (4000 bilhetes)
✓ 2026-01-19 (1000 bilhetes)
✓ 2026-01-20 (1000 bilhetes)
✓ 2026-01-21 (1000 bilhetes)
✓ 2026-01-22 (1000 bilhetes)
✓ 2026-01-23 (1000 bilhetes)
✓ 2026-01-24 (4000 bilhetes)
✓ 2026-01-25 (4000 bilhetes)
✓ 2026-01-26 (1000 bilhetes)
✓ 2026-01-27 (1000 bilhetes)
✓ 2026-01-28 (1000 bilhetes)
✓ 2026-01-29 (1000 bilhetes)
✓ 2026-01-30 (1000 bilhetes)
✓ 2026-01-31 (4000 bilhetes)
✓ 2026-02-01 (4000 bilhetes)
✓ 2026-02-02 (1000 bilhetes)
✓ 2026-02-03 (10

## PART II – Sistema de Informação e Aplicação

#### 3. Desenvolvimento de Aplicação [4 valores]

##### Critérios de Avaliação

* Funcionalidade dos *endpoints*
* Uso correto de transações
* Prevenção de injeção de SQL
* Uso correto de métodos e códigos de erro HTTP

`Nota` Todo o código da aplicação deve ser incluído na entrega do projeto.

`Nota` A aplicação deve ainda estar disponível *online* para demonstração durante a discussão oral, no ambiente de desenvolvimento Docker providenciado para a disciplina.

Crie um protótipo de RESTful *web service* para venda de bilhetes e votação nos recintos por acesso programático à base de dados ‘zoo’ através de uma API que devolve respostas em JSON, análoga à demonstrada no Lab10. A API deve implementar os seguintes *endpoints REST*:

| Endpoint | Descrição |
| ----- | ----- |
| /zona/\<zona\>/ | OUTPUT: lista de todos os ***recintos*** da \<zona\>, contendo o *número* do ***recinto*** e as ***espécies*** nele contidas, indicando, para cada ***espécie***, os *nomes científico* e *comum*, e o número de ***animais*** da ***espécie*** que estão no ***recinto***. |
| /recinto/\<recinto\>/voto/\<bilhete\>/ | Assinala o voto do \<bilhete\> no \<recinto\>, atualizando as tabelas ***bilhete*** (marcando *votou* TRUE) e ***recinto*** (incrementando os *votos*). Devolve mensagem de erro informativa e diferenciada (e não assinala o voto) se o \<bilhete\> já tinha *votou* \= TRUE ou se o \<recinto\> não está contido numa zona a que o \<bilhete\> tinha acesso. |
| /venda/ | Executa  uma venda de um ou mais bilhetes, ***populando*** as tabelas ***venda***, ***bilhete*** e ***acesso***. INPUT: NIF do cliente \[opcional\] mais lista de bilhetes, em que cada bilhete consiste numa lista de zonas de acesso, e um desconto \[opcional\]. OUTPUT: O preço total da venda, e a lista dos bilhetes da venda com o número e preço de cada um.  |

A solução deve prezar pela segurança, **prevenindo** ataques por **injeção de** **SQL**, e garantindo que as **operações de escrita** estão encapsuladas em **transações** que cumprem os princípios ACID.

0. Itemize quais as estratégias que utilizou:

* Estratégia 1: Lorem ipsum lorem ipsum lorem ipsum
* Estratégia 2: Lorem ipsum lorem ipsum lorem ipsum

1. Copie para a célula abaixo a função do app.py zona_index

`Nota` Formate o seu código previamente e mantenha o bloco ```python para colorizar a sintaxe

```python
@app.route("/", methods=("GET",))
@app.route("/accounts", methods=("GET",))
@limiter.limit("1 per second")
def account_index():
    """Show all the accounts, most recent first."""

    with pool.connection() as conn:
        with conn.cursor() as cur:
            accounts = cur.execute(
                """
                SELECT account_number, branch_name, balance
                FROM account
                ORDER BY account_number DESC;
                """,
                {},
            ).fetchall()
            log.debug(f"Found {cur.rowcount} rows.")

    return jsonify(accounts), 200
```

2. Copie para a célula abaixo a função do app.py recinto_voto_save

`Nota` Formate o seu código previamente e mantenha o bloco ```python para colorizar a sintaxe

```python
@app.route(
    "/accounts/<account_number>/update",
    methods=(
        "PUT",
        "POST",
    ),
)
def account_update_save(account_number):
    """Update the account balance."""

    balance = request.args.get("balance")

    error = None

    if not balance:
        error = "Balance is required."
    if not is_decimal(balance):
        error = "Balance is required to be decimal."

    if error is not None:
        return jsonify({"message": error, "status": "error"}), 400
    else:
        with pool.connection() as conn:
            with conn.cursor() as cur:
                cur.execute(
                    """
                    UPDATE account
                    SET balance = %(balance)s
                    WHERE account_number = %(account_number)s;
                    """,
                    {"account_number": account_number, "balance": balance},
                )
                # The result of this statement is persisted immediately by the database
                # because the connection is in autocommit mode.
                log.debug(f"Updated {cur.rowcount} rows.")

                if cur.rowcount == 0:
                    return (
                        jsonify({"message": "Account not found.", "status": "error"}),
                        404,
                    )

        # The connection is returned to the pool at the end of the `connection()` context but,
        # because it is not in a transaction state, no COMMIT is executed.

        return "", 204
```

3. Copie para a célula abaixo a função do app.py venda_save

`Nota` Formate o seu código previamente e mantenha o bloco ```python para colorizar a sintaxe

```python
@app.route(
    "/accounts/<account_number>/update",
    methods=(
        "PUT",
        "POST",
    ),
)
def account_update_save(account_number):
    """Update the account balance."""

    balance = request.args.get("balance")

    error = None

    if not balance:
        error = "Balance is required."
    if not is_decimal(balance):
        error = "Balance is required to be decimal."

    if error is not None:
        return jsonify({"message": error, "status": "error"}), 400
    else:
        with pool.connection() as conn:
            with conn.cursor() as cur:
                cur.execute(
                    """
                    UPDATE account
                    SET balance = %(balance)s
                    WHERE account_number = %(account_number)s;
                    """,
                    {"account_number": account_number, "balance": balance},
                )
                # The result of this statement is persisted immediately by the database
                # because the connection is in autocommit mode.
                log.debug(f"Updated {cur.rowcount} rows.")

                if cur.rowcount == 0:
                    return (
                        jsonify({"message": "Account not found.", "status": "error"}),
                        404,
                    )

        # The connection is returned to the pool at the end of the `connection()` context but,
        # because it is not in a transaction state, no COMMIT is executed.

        return "", 204
```

## PART III – Análise de Dados

A resolução das alíneas que se seguem **tem de obedecer às seguintes restrições**:
- Só pode haver **um único comando exterior** (CREATE, ALTER, UPDATE ou SELECT) e, no caso de envolver o comando SELECT, **não pode haver mais do que 1 nível de encadeamento de SELECTs**.
- **Não pode** recorrer aos operadores LIMIT ou WITH, ou a qualquer operador não coberto nos materiais da disciplina.


#### 4. Engenharia de Dados [2 valores]

##### Critérios de Avaliação

* Correção das soluções
* Eficiência das soluções
* Simplicidade das soluções

1. Crie uma vista materializada para auxiliar a direção do zoo a analisar as vendas de bilhetes e receitas do zoo e suas zonas, com o seguinte esquema:

    *vendas_zoo(bid, id_zona, mes, dia_da_semana, data, receita)*

    Em que teremos, para cada bilhete (*bid*) e zona (*id_zona*) a que o bilhete teve acesso, a data de venda do bilhete (e atributos derivados mes e dia_da_semana) e a receita da zona com o bilhete (i.e., o preço base da zona modificado pelo desconto).

In [14]:
%%sql zoo
DROP MATERIALIZED VIEW IF EXISTS vendas_zoo;
CREATE MATERIALIZED VIEW vendas_zoo AS
SELECT 
    b.bid AS bid, 
    a.id_zona AS id_zona, 
    EXTRACT(MONTH FROM v.data_hora) AS mes,
    EXTRACT(DOW FROM v.data_hora) AS dia_da_semana,
    CAST(v.data_hora AS DATE) AS data,
    --O arredondamento é apenas para ficarmos com duas casas decimais em vez de 20
    ROUND((1 - (b.desconto/100)) * z.preco, 2) AS receita 
FROM venda v
JOIN bilhete b ON v.no_venda = b.no_venda 
JOIN acesso a ON b.bid = a.bid 
JOIN zona z ON a.id_zona = z.id_zona;

++
||
++
++

In [15]:
%%sql zoo
SELECT * FROM vendas_zoo LIMIT 5;

bid,id_zona,mes,dia_da_semana,data,receita
50,1,1,4,2026-01-01,15.00
100,1,1,4,2026-01-01,7.50
150,1,1,4,2026-01-01,7.50
200,1,1,4,2026-01-01,15.00
250,1,1,4,2026-01-01,15.00


2. Modifique a tabela recinto adicionando um campo rentabilidade de tipo REAL

In [16]:
%%sql zoo
ALTER TABLE recinto
ADD COLUMN rentabilidade REAL;

++
||
++
++

3. Actualize a tabela recinto com o valor da rentabilidade sendo obtido pela receita total da zona onde está o recinto multiplicada pela fração entre os votos do recinto e o total de votos na zona. Pode recorrer à vista votos_zoo para realizar esta actualização.

In [20]:
%%sql zoo
UPDATE recinto r
SET rentabilidade = (SELECT SUM(receita) FROM vendas_zoo vz WHERE vz.id_zona = r.id_zona) * (r.votos * 1.0 / (
    SELECT SUM(votos) FROM recinto r2 WHERE id_zona = r2.id_zona
))

++
||
++
++

#### 5. Consultas [4 valores]

##### Critérios de Avaliação

* Correção das soluções
* Eficiência das soluções
* Simplicidade das soluções

Apresente a consulta SQL mais sucinta para cada um dos seguintes objetivos analíticos da empresa. Deve usar a vista vendas_zoo (exclusivamente ou em adição às outras tabelas) sempre que isso simplificar a consulta. Pode também usar operadores OLAP onde for adequado ao pedido.

1. Determinar o recinto mais rentável para cada zona, incluindo o valor da sua rentabilidade.

In [21]:
%%sql zoo
SELECT id_zona, r.id_recinto, r.rentabilidade
FROM recinto r
WHERE r.rentabilidade = (
    SELECT MAX(sub.rentabilidade)
    FROM recinto sub
    WHERE sub.id_zona = r.id_zona
);

id_zona,id_recinto,rentabilidade
2,16,19911.287
2,17,19911.287
2,18,19911.287
2,19,19911.287
2,20,19911.287
2,21,19911.287
2,22,19911.287
2,23,19911.287
2,24,19911.287
2,25,19911.287


2. Determinar se a aposta do zoo na sua especialidade está a compensar em termos de receitas, bilhetes ou votos, calculando as médias por zona, entre zonas da especialidade e zonas que não são da especialidade, de receitas globais, de bilhetes vendidos, e de votos recebidos.

In [2]:
%%sql zoo
SELECT
    CASE WHEN z.continente = 'África' THEN 'especialidade' ELSE 'outras' END AS tipo,
    ROUND(AVG((SELECT SUM(v.receita) FROM vendas_zoo v WHERE v.id_zona = z.id_zona)), 2) AS media_receita,
    ROUND(AVG((SELECT COUNT(*)       FROM vendas_zoo v WHERE v.id_zona = z.id_zona)), 2) AS media_bilhetes,
    ROUND(AVG((SELECT SUM(r.votos)   FROM recinto    r WHERE r.id_zona = z.id_zona)), 2) AS media_votos
FROM zona z
GROUP BY tipo;

tipo,media_receita,media_bilhetes,media_votos
especialidade,2157496.50,215236.00,34312.00
outras,2219349.20,213956.60,34275.00


3. Determinar a percentagem de bilhetes com acesso a todas as zonas, globalmente e com *drill downs* independentes por dia da semana e por mês.

In [23]:
%%sql zoo
SELECT 
    mes,
    dia_da_semana,
    ROUND(COUNT(DISTINCT bid) * 100.0 / 
           (SELECT COUNT(DISTINCT v2.bid)
            FROM vendas_zoo v2
            WHERE (v1.mes IS NULL OR v2.mes = v1.mes)
              AND (v1.dia_da_semana IS NULL OR v2.dia_da_semana = v1.dia_da_semana)), 2) AS percentagem
FROM (
    SELECT DISTINCT
        v1.bid,
        v1.mes,
        v1.dia_da_semana
    FROM vendas_zoo v1
    WHERE NOT EXISTS (
        SELECT id_zona FROM zona
        EXCEPT
        SELECT a.id_zona 
        FROM acesso a 
        WHERE a.bid = v1.bid
    )
) AS v1
GROUP BY ROLLUP(mes, dia_da_semana);

mes,dia_da_semana,percentagem
1,0,2.00
1,1,26.50
1,2,2.00
1,3,26.50
1,4,2.00
1,5,2.00
1,6,21.60
1,None,12.14
2,0,26.50
2,1,2.00


4. Determinar que zonas podem ter um aumento no preço e que zonas devem ter uma redução no preço, analisando a média diária de bilhetes vendidos, globalmente e com vértices nas dimensões zona e mês.

In [24]:
%%sql zoo
SELECT 
    id_zona,
    mes,
    ROUND(COUNT(DISTINCT bid) * 1.0 / COUNT(DISTINCT data), 2) AS media_diaria_bilhetes
FROM vendas_zoo
GROUP BY CUBE(id_zona, mes);

id_zona,mes,media_diaria_bilhetes
1,1,1460.00
1,2,1052.14
1,3,1523.23
1,4,1538.67
1,5,1461.94
1,6,1456.36
1,None,1416.30
2,1,1523.23
2,2,1472.14
2,3,1207.10


#### 6. Índices [3 valores]

##### Critérios de Avaliação

* Utilidade dos índices
* Adequação  da justificação

É expectável que seja necessário executar consultas semelhantes **às consultas 1 e 2 do ponto anterior** diversas vezes ao longo do tempo, pelo que pretendemos otimizar o desempenho dessas consultas por meio da criação de índices. Crie sobre a vista vendas\_zoo ou sobre as restantes tabelas da base de dados o(s) índice(s) que achar mais indicados para fazer essa otimização, justificando a sua escolha com **argumentos teóricos** e com **demonstração prática** do ganho em eficiência do índice por meio do comando EXPLAIN ANALYSE. Deve procurar uma otimização coletiva das duas consultas, evitando criar índices excessivos, particularmente se estes trazem apenas ganhos incrementais. Nomeadamente, devido a preocupações com o armazenamento, não se considera vantajoso atingir planos de execução com INDEX ONLY SCAN se isso obriga a colocar no índice mais atributos do que estritamente necessário para conseguir um INDEX SCAN.

##### Consulta 1

1. Demonstração do custo (com EXPLAIN ANALYSE)

In [3]:
%%sql zoo
DROP INDEX IF EXISTS idx_recinto_zona_rent;
EXPLAIN ANALYZE
SELECT id_zona, r.id_recinto, r.rentabilidade
FROM recinto r
WHERE r.rentabilidade = (
    SELECT MAX(sub.rentabilidade)
    FROM recinto sub
    WHERE sub.id_zona = r.id_zona
);

QUERY PLAN
Seq Scan on recinto r (cost=0.00..356.11 rows=1 width=12) (actual time=0.372..2.332 rows=91.00 loops=1)
Filter: (rentabilidade = (SubPlan 1))
Rows Removed by Filter: 14
Buffers: shared hit=212
SubPlan 1
-> Aggregate (cost=3.35..3.36 rows=1 width=4) (actual time=0.021..0.021 rows=1.00 loops=105)
Buffers: shared hit=210
-> Seq Scan on recinto sub (cost=0.00..3.31 rows=15 width=4) (actual time=0.008..0.017 rows=15.00 loops=105)
Filter: (id_zona = r.id_zona)
Rows Removed by Filter: 90


2. Comandos de criação do(s) índice(s)

In [3]:
%%sql zoo
CREATE INDEX IF NOT EXISTS idx_recinto_zona_rent ON recinto USING BTREE (id_zona, rentabilidade);

++
||
++
++

3. Demonstração do custo final (com EXPLAIN ANALYSE)

In [4]:
%%sql zoo
EXPLAIN ANALYZE
SELECT id_zona, r.id_recinto, r.rentabilidade
FROM recinto r
WHERE r.rentabilidade = (
    SELECT MAX(sub.rentabilidade)
    FROM recinto sub
    WHERE sub.id_zona = r.id_zona
);

QUERY PLAN
Seq Scan on recinto r (cost=0.00..356.11 rows=1 width=12) (actual time=0.336..1.857 rows=91.00 loops=1)
Filter: (rentabilidade = (SubPlan 1))
Rows Removed by Filter: 14
Buffers: shared hit=212
SubPlan 1
-> Aggregate (cost=3.35..3.36 rows=1 width=4) (actual time=0.016..0.017 rows=1.00 loops=105)
Buffers: shared hit=210
-> Seq Scan on recinto sub (cost=0.00..3.31 rows=15 width=4) (actual time=0.006..0.013 rows=15.00 loops=105)
Filter: (id_zona = r.id_zona)
Rows Removed by Filter: 90


4. Justificação

A consulta executa um auto-join sobre a tabela recinto para identificar os recintos que atingem a rentabilidade máxima em cada zona. Sem índice, o otimizador é forçado a realizar dois Seq Scan completos à tabela.

O índice composto (`id_zona`, `rentabilidade`) mitiga este problema em duas fases:

1. **Subconsulta sobre recinto**: A tabela tem 105 linhas. Sem optimização, cada linha seria comparada com todas as outras, O(N²), com N = 105. O indice permite a filtragem por zona, o que reduz de O(N²) para O(N*M), onde M é o número de recintos com o mesmo `id_zona`. Ao estar agregado à `rentabilidade`, este indice também permite, para além de organizar as linhas por `id_zona`, organizá-las por `rentabilidade`, dentro da zona, permitindo um `MAX` mais rápido, que apenas não se mostrou porque todos os valores são iguais, logo tem à mesma de percorrer a tabela, mas em maior parte dos casos, terá grandes ganhos. Para além disso, faz com que a rentabilidade seja acedida no indice, removendo a necessidade de Heap Fetches.
O outer Seq Scan mantém-se, para percorrer cada recinto. Como percorre todos os recintos, não teria vantagem em usar indice, logo não foi criado um indice para `id_recinto`.

Estas optimizações fazem com que o tempo de execução reduza significativamente, de ~0.6ms para ~0.2ms. 

##### Consulta 2

1. Demonstração do custo inicial (com EXPLAIN ANALYSE)

In [33]:
%%sql zoo
DROP INDEX IF EXISTS idx_recinto_zona_votos;
DROP INDEX IF EXISTS idx_vendas_zoo_zona_receita;

EXPLAIN ANALYSE
SELECT
    CASE WHEN z.continente = 'África' THEN 'especialidade' ELSE 'outras' END AS tipo,
    ROUND(AVG((SELECT SUM(v.receita) FROM vendas_zoo v WHERE v.id_zona = z.id_zona)), 2) AS media_receita,
    ROUND(AVG((SELECT COUNT(*)       FROM vendas_zoo v WHERE v.id_zona = z.id_zona)), 2) AS media_bilhetes,
    ROUND(AVG((SELECT SUM(r.votos)   FROM recinto    r WHERE r.id_zona = z.id_zona)), 2) AS media_votos
FROM zona z
GROUP BY tipo;

QUERY PLAN
GroupAggregate (cost=112.96..95338965.06 rows=200 width=128) (actual time=943.817..1760.418 rows=2.00 loops=1)
Group Key: (CASE WHEN (z.continente = 'África'::cnt) THEN 'especialidade'::text ELSE 'outras'::text END)
Buffers: shared hit=155023
-> Sort (cost=112.96..116.89 rows=1570 width=36) (actual time=265.518..265.527 rows=7.00 loops=1)
Sort Key: (CASE WHEN (z.continente = 'África'::cnt) THEN 'especialidade'::text ELSE 'outras'::text END)
Sort Method: quicksort Memory: 25kB
Buffers: shared hit=1
-> Seq Scan on zona z (cost=0.00..29.63 rows=1570 width=36) (actual time=265.443..265.449 rows=7.00 loops=1)
Buffers: shared hit=1
SubPlan 1


2. Comandos de criação do(s) índice(s)

In [34]:
%%sql zoo
CREATE INDEX IF NOT EXISTS idx_vendas_zoo_zona_receita ON vendas_zoo USING BTREE (id_zona, receita);
CREATE INDEX IF NOT EXISTS idx_recinto_zona_votos ON recinto (id_zona, votos);

++
||
++
++

3. Demonstração do custo final (com EXPLAIN ANALYSE)

In [35]:
%%sql zoo

EXPLAIN ANALYSE
SELECT
    CASE WHEN z.continente = 'África' THEN 'especialidade' ELSE 'outras' END AS tipo,
    ROUND(AVG((SELECT SUM(v.receita) FROM vendas_zoo v WHERE v.id_zona = z.id_zona)), 2) AS media_receita,
    ROUND(AVG((SELECT COUNT(*)       FROM vendas_zoo v WHERE v.id_zona = z.id_zona)), 2) AS media_bilhetes,
    ROUND(AVG((SELECT SUM(r.votos)   FROM recinto    r WHERE r.id_zona = z.id_zona)), 2) AS media_votos
FROM zona z
GROUP BY tipo;

QUERY PLAN
GroupAggregate (cost=112.96..24362624.56 rows=200 width=128) (actual time=241.146..602.271 rows=2.00 loops=1)
Group Key: (CASE WHEN (z.continente = 'África'::cnt) THEN 'especialidade'::text ELSE 'outras'::text END)
Buffers: shared hit=5798 read=5757
-> Sort (cost=112.96..116.89 rows=1570 width=36) (actual time=113.888..113.901 rows=7.00 loops=1)
Sort Key: (CASE WHEN (z.continente = 'África'::cnt) THEN 'especialidade'::text ELSE 'outras'::text END)
Sort Method: quicksort Memory: 25kB
Buffers: shared hit=1
-> Seq Scan on zona z (cost=0.00..29.63 rows=1570 width=36) (actual time=113.862..113.868 rows=7.00 loops=1)
Buffers: shared hit=1
SubPlan 1


4. Justificação

Os índice `idx_vendas_zoo_zona_receita` foi usado nesta consulta, mas o `idx_recinto_zona_votos` não foi usado, como demosntra o EXPLAIN ANALYSE.

1. **Subconsulta sobre `recinto`**: A tabela contém apenas 105 linhas. Como esta tabela tem dimensão reduzida, o custo de de um B-Tree supera o custo do Seq Scan, razão pela qual este ultimo é escolhido.

2. **Subconsulta sobre `vendas_zoo`**: Esta é uma tabela maior, e a consulta possui uma cláusula que restrinje o conjunto de linhas a processar, logo, para cada subcunsulta, apenas são processadas as linhas que verificam id_zona = z.id_zona, reduzindo o número total de linhas processadas. para além disso, como o indice contém `id_zona` e `receita`, não é necessário aceder ao heap para aceder à receita associada à linha, porque obtemos o seu valor através do indice.

Concluindo, nesta consulta, a utilização de indices permitiu reduzir bastante significativamente o tempo de procura, de ~2000ms para ~600ms.

#### Entrega

A submissão do projeto deve ser feita na forma de um arquivo zip de nome **entrega-bd-02-GG.zip**, onde **GG** é o número do grupo, estruturado da seguinte forma:

| Ficheiro | Descrição |
| :---- | :---- |
| entrega-bd-02-GG.ipynb | Um ficheiro Jupyter Notebook correspondendo ao preenchimento do template “entrega-bd-02-GG.ipynb” disponibilizado na página da disciplina (onde GG deverá ser subtituído pelo número do grupo).<br/><br/>Deverá preencher a primeira célula com o número do grupo, o número e nome dos alunos que o constituem, tal como a percentagem relativa de contribuição de cada aluno com o respectivo esforço (horas).<br/><br/>Deverá ainda preencher no Jupyter Notebook as respostas a todas as perguntas para as quais há um campo de resposta.<br/><br/>O ficheiro “entrega-bd-02-GG.ipynb” pode ser importado para o ambiente de trabalho disponibilizado para as aulas de laboratório (i.e., https://github.com/bdist/bdist-workspace), que serve de ambiente de teste para as partes em SQL.<br/><br/>Deve certificar-se que todo o código SQL é executável no ambiente de trabalho, |
| app/ | Pasta com os ficheiros que compõem a aplicação web correspondendo à secção<br/><br/>3\. **Desenvolvimento de Aplicação** |

**IMPORTANTE**

* Serão aplicadas penalizações aos grupos que não cumprirem o formato de submissão.
* Não serão aceites submissões fora do prazo.